Add support for partial http requests handling with pipelining

mhl-b opened the pull request:
UPDATE: 
Created a leaner and cleaner version here https://github.com/elastic/elasticsearch/pull/111438. 
Most of changes here are irrelevant in new PR.


-----

This PR adds support to `Netty4HttpPipeliningHandler` to handle parts of HTTP request. Right now we always aggregate HTTP request content before passing to `RestController`. With this change `Netty4HttpPipeliningHandler` can receive `FullHttpRequests` and `HttpRequests`/`HttpContents` with correct sequence number.

To make it possible I made several changes in Netty channel pipeline.
1. Add HTTP pipeline sequence to parts of HTTP request. Before `Netty4HttpPipeliningHandler` would keep track of all incoming `FullHttpRequests` and increase sequence number. With this change addition of sequence number happens sooner, after `HttpRequestDecoder`, but before decompression and aggregation. It happens in `Netty4InboundHttpPipeliningHandler`.

2. Propagation of sequence number through `HttpObjectAggregator` and `HttpContentDecompressor`. Since netty does not know about our pipelining implementation, I have to wrap both classes that can consume and produce a "pipelined" versions of `HttpObjects`. Besides wrapper handler classes I introduced strong-typed versions of `HttpRequest, HttpContent, LastHttpContent, FullHttpRequest` with pipeline sequence number, for example 
`public record PipelinedHttpContent(HttpContent httpContent, int sequence)`
3. Make HTTP aggregation optional. In this PR we still do full aggregation to all requests. But I added a predicate that can control which requests can skip aggregation, and unit tests.

After change pipeline looks like this:
```
| from                        | to                                 | note                       |
|-----------------------------+------------------------------------+----------------------------|
| HttpRequestDecoder          | HttpRequestDecoder                 |                            |
| Netty4HttpHeaderValidator   | Netty4HttpHeaderValidator          |                            |
|                             | Netty4InboundHttpPipeliningHandler | emits PipelinedHttpObjects |
| HttpContentDecompressor     | Netty4ContentDecompressor          | wrapped decompressor       |
| HttpObjectAggregator        | Netty4HttpAggregator               | optional aggregation       |
| Netty4HttpPipeliningHandler | Netty4HttpPipeliningHandler        |                            |
```

Most interesting changes in `Netty4ContentDecompressor, Netty4HttpAggregator, Netty4InboundHttpPipeliningHandler , Netty4HttpPipeliningHandler` and related unit tests at the end. The rest is boilerplate.

**Update**:
Added example of RestHandler with chunked content. I use [reactive streams model](https://www.reactive-streams.org/). Rather than exposing `BytesReference` as content, I use `Flow.Publisher<HttpContent>`. The RestHandler will subscribe to publisher and consume parts at it's own rate - by [`Subscription.request(long n)`](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/util/concurrent/Flow.Subscription.html#request(long)). The publisher will read from netty channel on demand and invoke `onNext(HttpContent)` when number of requested parts is greater than 0.

Content publisher - `Netty4RequestContentPublisher`
Rest handler example - `Netty4RequestContentPublisherIT`

I omitted error handling and edge cases for brevity. There many things to consider with publisher state machine: for example avoid multiple subscriptions, handling terminal states, handling errors, handling subscriber cancellation. I did some drafts, and I can say it's pretty easy to add them.

mhl-b committed:
add support for partial http requests with pipelining

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

mhl-b commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:
```diff
@@ -109,28 +107,40 @@ public Netty4HttpPipeliningHandler(
         }
     }
 
+    private void handlePipelinedFullRequest(ChannelHandlerContext ctx, PipelinedFullHttpRequest request) {
+        final Netty4HttpRequest netty4HttpRequest;
+        if (request.decoderResult().isFailure()) {
+            final Throwable cause = request.decoderResult().cause();
+            final Exception nonError;
+            if (cause instanceof Error) {
+                ExceptionsHelper.maybeDieOnAnotherThread(cause);
+                nonError = new Exception(cause);
+            } else {
+                nonError = (Exception) cause;
+            }
+            netty4HttpRequest = new Netty4HttpRequest(request.sequence(), request, nonError);
+        } else {
+            netty4HttpRequest = new Netty4HttpRequest(request.sequence(), request);
+        }
+        handlePipelinedRequest(ctx, netty4HttpRequest);
+    }
```
no changes here, moved into separate method

DaveCTurner reviewed (commented):
Thanks Mikhail. I'll take a more detailed look in due course but two things straight away:

1. Could we have a test case which shows how a REST handler would use this API? Probably a `ESNetty4IntegTestCase` derivative would be best, see e.g. `Netty4ChunkedContinuationsIT` for an example of the sort of thing I mean.

2. Could we merge this into a feature branch for now rather than `main`? That way we've somewhere to coordinate the work to adopt this in the bulk indexing path, but we're not committing to it until we have the whole solution in place and can run some benchmarks?

mhl-b commented:
@DaveCTurner 
1. Make sense, will do
2. Feature branch - done

ywangd commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/PipelinedHttpRequestPart.java:
```diff
@@ -0,0 +1,17 @@
+/**
+ * A part of HTTP request that has HTTP pipelining sequence number
+ */
+public sealed interface PipelinedHttpRequestPart extends PipelinedHttpObject permits PipelinedHttpRequest, PipelinedHttpContent,
+    PipelinedLastHttpContent {
```
I'd prefer to drop this marker interface. IIUC, it is only to differentiate `PipelinedFullHttpRequest` from other `PipelinedHttpObject`. Can we instead have a method on `PipelinedHttpObject` to differentiate between them? I personally find the hierarchy a bit hard to follow.

mhl-b replied:
I planned to use this sealed interface in PartialRequest path, so I can do `switch(part)` with exhaustive check. Which is part of java 21 [jep-441](https://openjdk.org/jeps/441). But we are still compiling with java 17 and nothing fancy came out of this code. Can be removed.

ywangd reviewed (commented):
This seems to be an exciting piece of work! Thanks for working on it. I had a first read-through and left some minor comments. I plan to come back to it and have a closer look. In the meantime, I second on David comment about having a test case that demonstrates how this new feature is leveraged.

mhl-b commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4ContentDecompressor.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public boolean acceptInboundMessage(Object msg) throws Exception {
+        return msg instanceof PipelinedHttpObject;
```
Wrapped Decompressor should call this method on every "read(ctx, msg)", I added this check to avoid surprises in "decode" method that assumes message has sequence number.

mhl-b committed:
add rest handler example

mhl-b commented:
@DaveCTurner @ywangd 
I updated PR description with approach and implementation of RestHandler with chunked request content.
Please see `Netty4RequestContentPublisherIT` for an API design. TLDR I use [reactive streams](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/util/concurrent/Flow.html).

mhl-b commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpAggregator.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public class Netty4HttpAggregator extends HttpObjectAggregator {
+
+    private static final Predicate<PipelinedHttpRequest> IGNORE_TEST = (req) -> req.uri().startsWith("/_test/request-stream") == false;
```
added to the source for the sake of PR size

mhl-b commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:
```diff
@@ -109,23 +112,34 @@ public Netty4HttpPipeliningHandler(
+                    } else {
+                        var contentPublisher = new Netty4RequestContentPublisher(ctx.channel());
+                        netty4HttpRequest = new Netty4HttpRequest(sequence, request, contentPublisher);
+                        currentRequest = netty4HttpRequest;
```
note: I attach current request to Handler object and following chunks will go to the same object until we see next request

mhl-b commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:
```diff
@@ -109,23 +112,34 @@ public Netty4HttpPipeliningHandler(
             } else {
-                netty4HttpRequest = new Netty4HttpRequest(readSequence++, fullHttpRequest);
+                assert currentRequest != null;
+                currentRequest.contentPublisher().sendChunk((HttpContent) msg);
```
I attach currentRequest to the Netty4PipeliningHandler and consequent chunks can be published right a way to the RestHandler from here.

mhl-b commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java:
```diff
@@ -36,39 +42,64 @@
     private final boolean pooled;
     private final int sequence;
 
+    Netty4HttpRequest(int sequence, HttpRequest request, Netty4RequestContentPublisher contentPublisher) {
+        this.sequence = sequence;
+        this.nettyRequest = request;
+        this.nettyContent = new DefaultHttpContent(Unpooled.EMPTY_BUFFER);
+        this.contentPublisher = contentPublisher;
+        this.content = BytesArray.EMPTY;
+        this.released = new AtomicBoolean(false);
+        this.pooled = false;
+        this.headers = getHttpHeadersAsMap(request.headers());
+        this.inboundException = null;
+    }
```
Add support to Netty4HttpRequest to be HttpRequest and optional HttpContent, rather than always FullHttpRequest. I do drop trailing headers for not full requests, not sure do we use them or not. Haven't found traces of them yet.

mhl-b commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4RequestContentPublisher.java:
```diff
@@ -0,0 +1,60 @@
+
+import java.util.concurrent.Flow;
+
+public class Netty4RequestContentPublisher implements Flow.Publisher<HttpContent> {
```
bare bones, nothing fancy, no error handling, no state management, no multiple subscriptions, just happy path

DaveCTurner reviewed (commented):
A reactive stream is the right conceptual model for sure but I'm not convinced we want to stick to these exact APIs. There's always exactly one subscriber, known ahead of time, so there's no need for a separate `subscribe()` API; we really want `.request()` to request a certain number of bytes but will typically deliver this as a single chunk, which differs from the reactive streams API that expects the argument to be the number of chunks; finally the `Throwable` in the `onError` path is very concerning. Remember this code is all on an incredibly hot path, we need to avoid unnecessary abstractions wherever possible.

mhl-b commented:
@DaveCTurner @ywangd 
I created new PR https://github.com/elastic/elasticsearch/pull/111438. 
I found simpler and smaller implementation, that does not have problems introduced here. No changes in pipelining sequencing, no more PipelinedObjects, simplified wrapper for HttpObjectAggregator that does not intercept "decode" method, simplified interface for streamed content, I still use reactive streams as a model but only with minimal required methods.

ywangd commented:
Would it be OK to close this PR if this is not intended to be continued?
